explorer/interpreter/interpreter.cpp

```diff
@@ -182,7 +184,15 @@ class Interpreter {
       -> ErrorOr<Success>;
 
   auto CallDestructor(Nonnull<const DestructorDeclaration*> fun,
-                      Nonnull<const Value*> receiver) -> ErrorOr<Success>;
+                      Nonnull<const Value*> receiver,
+                      const LocationValue* class_addr) -> ErrorOr<Success>;
```

Comment:
This method should have a comment. I'm particularly looking to understand the meaning of `class_addr`.

Reply:
I will happily write a comment and fix the comment on the other method, and was hoping for some feedback about this approach.

The problem is within `BindSelfIfPresent`. If the source code is using the immutable `[self: Self]` pattern, which is the only binding destructors used to support, the `method_scope.BindValue(` call would happily bind to the `NominalClassValue` provided by calling code in `receiver`. However, in the `[addr self: Self*]` case, `PatternMatch` will assert on a binding to any value that is not a `LocationValue` type. So I now ask the caller to provide both values.

I could refactor the code in the interpreter that calls `CallDestructor` to only provide one, but it would then need to know the arguments to the destructor to understand which one to provide, thus duplicating the logic in `BindSelfIfPresent`. So I've gone with the 2 args. I'd love if there's a more elegant way to do this, do you have anything in mind?

Reply:
I think that, rather than construct an `ExpressionResult` on the fly for the `PatternMatch` call, we should pass the receiver into `CallFunction` and `CallDestructor` as an `ExpressionResult`. This seems like exactly the kind of problem that `ExpressionResult` is designed to solve: representing a value that might or might not have an address, in a context that might want to operate on the value or the address. As a bonus, I think this could also help resolve the TODO in `BindSelfIfPresent`.

(I'm a little concerned about the way `ExpressionResult` blurs the line between compile-time and run-time, but that's a pre-existing problem, and a much bigger one. I have a terrible feeling the answer there will wind up being that we have to split `PatternMatch` up into `Action`s, so that it can do things like evaluate the implicit address-of expression in an `addr` pattern.)

Reply:
Done, PTAL